server/datastore/mysql/microsoft_mdm.go

```diff
@@ -1306,44 +1307,73 @@ func listMDMWindowsProfilesToInstallDB(
-		if len(onlyProfileUUIDs) > 0 {
-			query, args, err = sqlx.In(
-				query,
-				onlyProfileUUIDs, hostUUIDs,
-				onlyProfileUUIDs, hostUUIDs,
-				onlyProfileUUIDs, hostUUIDs,
-				onlyProfileUUIDs, hostUUIDs,
-				fleet.MDMOperationTypeInstall,
-			)
-		} else {
-			query, args, err = sqlx.In(query, hostUUIDs, hostUUIDs, hostUUIDs, hostUUIDs, fleet.MDMOperationTypeInstall)
+	toInstallQuery = fmt.Sprintf(toInstallQuery, hostFilter, hostFilter, hostFilter, hostFilter)
+
+	// use a 10k host batch size to match what we do on the macOS side.
+	selectProfilesBatchSize := 10_000
+	if ds.testSelectMDMProfilesBatchSize > 0 {
+		selectProfilesBatchSize = ds.testSelectMDMProfilesBatchSize
+	}
+	selectProfilesTotalBatches := int(math.Ceil(float64(len(hostUUIDs)) / float64(selectProfilesBatchSize)))
+
+	if len(hostUUIDs) == 0 {
+		// Make sure we run the query at least once
+		selectProfilesTotalBatches = 1
+	}
+
+	var wantedProfiles []*fleet.MDMWindowsProfilePayload
+	for i := range selectProfilesTotalBatches {
+		start := i * selectProfilesBatchSize
+		end := min(start+selectProfilesBatchSize, len(hostUUIDs))
+
+		batchUUIDs := hostUUIDs[start:end]
+
+		var err error
+		args := []any{fleet.MDMOperationTypeInstall}
+		stmt := toInstallQuery
+		if len(hostUUIDs) > 0 {
+			if len(onlyProfileUUIDs) > 0 {
+				stmt, args, err = sqlx.In(
+					toInstallQuery,
+					onlyProfileUUIDs, batchUUIDs,
+					onlyProfileUUIDs, batchUUIDs,
+					onlyProfileUUIDs, batchUUIDs,
+					onlyProfileUUIDs, batchUUIDs,
+					fleet.MDMOperationTypeInstall,
+				)
+			} else {
+				stmt, args, err = sqlx.In(toInstallQuery, batchUUIDs, batchUUIDs, batchUUIDs, batchUUIDs, fleet.MDMOperationTypeInstall)
+			}
+			if err != nil {
+				return nil, ctxerr.Wrapf(ctx, err, "building sqlx.In for list MDM windows profiles to install, batch %d of %d", i, selectProfilesTotalBatches)
+			}
 		}
+
+		var partialResult []*fleet.MDMWindowsProfilePayload
+		err = sqlx.SelectContext(ctx, tx, &partialResult, stmt, args...)
 		if err != nil {
-			return nil, ctxerr.Wrap(ctx, err, "building sqlx.In")
+			return nil, ctxerr.Wrapf(ctx, err, "selecting windows MDM profiles to install, batch %d of %d", i, selectProfilesTotalBatches)
 		}
+
+		wantedProfiles = append(wantedProfiles, partialResult...)
+
 	}
 
-	var profiles []*fleet.MDMWindowsProfilePayload
-	err = sqlx.SelectContext(ctx, tx, &profiles, query, args...)
-	return profiles, err
+	return wantedProfiles, nil
 }
 
 func (ds *Datastore) ListMDMWindowsProfilesToRemove(ctx context.Context) ([]*fleet.MDMWindowsProfilePayload, error) {
 	var result []*fleet.MDMWindowsProfilePayload
 	err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
 		var err error
-		result, err = listMDMWindowsProfilesToRemoveDB(ctx, tx, nil, nil)
+		result, err = ds.listMDMWindowsProfilesToRemoveDB(ctx, tx, nil, nil)
 		return err
 	})
 
 	return result, err
 }
 
-func listMDMWindowsProfilesToRemoveDB(
+func (ds *Datastore) listMDMWindowsProfilesToRemoveDB(
```

Comment:
Same feedback as for the install case. Feels even more like we should do this refactor given that doing so doesn't require us to modify the datastore interface.